doc/decisions/0_drafts/errors_core.md

```diff
@@ -278,12 +287,77 @@ if (found == NULL)
 
 ## Decision
 
-**Suggestion:** Go with "Pragmatic solution"
+Go with "Pragmatic solution".
+We don't define any strict rules for the entire API.
+Instead, we define a few basic guidelines.
+
+### Guidelines
+
+These guidelines shall be applied to each function of the public API separately to decide how the function treats errors.
+
+#### Boolean functions
+
+If a function is conceptually boolean, it should return a `bool`.
+This means choosing `true` or `false` as the fallback value used in case of errors.
+
+If choosing `true` or `false` is not feasible, redesigning the function may be necessary.
+The function may be doing to many checks in one go and splitting it into multiple functions could make sense.
+
+If splitting is also not possible, an `int` or `enum` should be returned.
+The name of such a function must make clear that it is not actually boolean (avoid `is` and `has`, prefer `check`) to avoid accidental use in an `if`.
+
+#### Check preconditions
+
+Functions must check all preconditions.
+A (bug-free) function should never segfault, abort the process or otherwise cause unexpected behavior on the caller side.
+All conditions that could cause such behavior must be listed as preconditions and checked by the function as a safety measure.
+
+#### `NULL` arguments
+
+A `NULL` pointer argument should never be treated as a separate error case (unless it is the only error case).
+If a `NULL` pointer constitutes an invalid argument, the function should return an existing error (might be used for other precondition checks) or a default fallback value.
```

Comment:
Unclear what this means. If we have such a problem, e.g. in the core, the decision should clearly point out the problems.

Reply:
What I wanted to say is: IMO you should not force yourself to invent an error value just for "illegal `NULL` argument". Try to handle it some other way, or combine it with other precondition failures. If "`NULL` argument" is the only error case and you already return an `int`, you can of course use an error value for `NULL`. But it should even at that point be clear, that the same error code will be used if additional preconditions are added in the future.

Some examples:
```c
//////// combines NULL with (potential) other errors
// returns 0 on success
// returns -1 on precondition violations, including key == NULL
int foo(Key * key);

/////// defines logic for NULL, i.e. NULL is not an invalid argument
// returns whether other is below key
// NULL is considered below no key, and no key is considered below NULL, i.e. returns false if at least one argument is NULL
bool keyIsBelow (Key * key, Key * other);

//////// uses default/fallback value for NULL
// returns namespace of key, for key == NULL the fallback value KEY_NS_NONE is returned
elektraNamespace keyGetNamespace(Key * key);
```

> the decision should clearly point out the problems.

Again, no concrete problems. Just a guideline to base APIs on.